### PR TITLE
message api endpoints

### PR DESCRIPTION
**Pull request by food0903**
adding the necessary BE endpoints for message api
* /send POST endpoint (with id, api key, and message as body params)
* /listen GET SSE endpoint (api key and id as query params) for any user to listen to any chatroom

in addition, we also include the modal in the user profile settings to either generate an api key or grab an api key if the api key already exists

### CODE REVIEW

**Comment by jonathanguven** [comment]
some changes nice

**Comment by food0903** [comment]
yup bro

**Comment by flotoria** [comment]
thats cool and dope

File: api/main_endpoints/routes/Messages.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      return;
+    }
+    if (result) {
+      apiKeyFound = true;
+    }
+  });
+
+  if(!apiKey || !message || !id){
+    res.sendStatus(FORBIDDEN);
+    return;
```
Comment (evanugarte) [inline api/main_endpoints/routes/Messages.js]: this if statement should go before the user.findOne, leave the else if where it is

Comment (flotoria) [reply api/main_endpoints/routes/Messages.js]: ryan is solving the issue rn lets gooooo!

File: src/Pages/Profile/MemberView/GetApiKeyModal.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import React, {useState, useEffect} from 'react';
+import { getApiKey } from '../../../APIFunctions/User';
+
+const GetApiKeyModal = (props) => {
+  const { bannerCallback = () => {} } = props;
+  const [apiKey, setApiKey] = useState('');
+
+  const INPUT_CLASS_NAME = 'indent-2 block w-full rounded-md border-0 py-1.5 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:  focus:ring-2 focus:ring-insetfocus:ring-indigo-600 sm:text-sm sm:leading-6 text-white';
```
Comment (akshtshrma24) [inline src/Pages/Profile/MemberView/GetApiKeyModal.js]: why is this here, if only used once but for some reason all others dont have their css defined

Comment (food0903) [reply src/Pages/Profile/MemberView/GetApiKeyModal.js]: fixeddddd!!!!!!!!!!!!!!!!!!!!!!

**Review by akshtshrma24 — APPROVED** [review]

File: api/main_endpoints/routes/Messages.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+router.post('/send', async (req, res) => {
+
+  const {apiKey, message, id} = req.body;
+
+  let apiKeyFound = false;
+
+  if(!apiKey || !message || !id){
+    res.sendStatus(FORBIDDEN);
+    return;
+  }
```
Comment (evanugarte) [inline api/main_endpoints/routes/Messages.js]: lets do this, leave a nice response. does it work?
```suggestion
  const required = [
    {value: apiKey, title: 'API Key', },
    {value: message, title: 'Message', },
    {value: id, title: 'Room ID', },
  ];
  const missingValue = required.find(({value}) => !value);
  if (missingValue){
    res.status(BAD_REQUEST).send(`You must specify a ${missingValue.title}`);
    return;
  }
```

File: api/main_endpoints/routes/Messages.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  await User.findOne({apiKey}, (error, result) => {
```
Comment (evanugarte) [inline api/main_endpoints/routes/Messages.js]: we dont need to await if we are returning a response to the client inside the findOne callback

File: api/main_endpoints/routes/Messages.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    if (error) {
+      res.sendStatus(SERVER_ERROR);
```
Comment (evanugarte) [inline api/main_endpoints/routes/Messages.js]: lets log the error, import logger, see other files for how to do this

File: api/main_endpoints/routes/Messages.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+      return;
+    }
+    if (result) {
+      apiKeyFound = true;
+    }
```
Comment (evanugarte) [inline api/main_endpoints/routes/Messages.js]:
```suggestion
      if (result) {
        writeMessage(id, message);
        return res.json({status: 'Message sent'});
     }
    return res.sendStatus(UNAUTHORIZED);
```

File: api/main_endpoints/routes/Messages.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  });
+
+  if (apiKeyFound === false) {
+    res.sendStatus(UNAUTHORIZED);
+    return;
+  }
+  writeMessage(id, message);
+  return res.json({status: 'Message sent'});
+
+});
```
Comment (evanugarte) [inline api/main_endpoints/routes/Messages.js]: put this in the callback instead, see other comment

File: api/main_endpoints/routes/Messages.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  try {
+    User.findOne({apiKey}, (error, result) => {
+      if (error) {
+        logger.error('/listen received an invalid API key: ', error);
+        res.sendStatus(SERVER_ERROR);
+        return;
+      }
+      if (result) { // logic to keep the connection alive
```
Comment (evanugarte) [inline api/main_endpoints/routes/Messages.js]: lets invert the logic for less indentation

```js
if (!result) {
  return res.sendStatus(UNAUTHORIZED);
}
// continue your code below, no need for an `else`
const headers = {
          'Content-Type': 'text/event-stream',
          'Connection': 'keep-alive',
          'Cache-Control': 'no-cache',
};
// ...
```

File: src/APIFunctions/User.js [hunk 9]
```diff
@@ -299,3 +299,14 @@ export async function getAllUsersValidVerifiedAndSubscribed(token) {
     });
   return status;
 }
+
+export async function getApiKey(token) {
+  let status = new UserApiResponse();
+  try {
+    const response = await axios.post(`${GENERAL_API_URL}/user/apikey`, { token });
```
Comment (evanugarte) [inline src/APIFunctions/User.js]:
```suggestion
    const url = new URL('/user/apikey', GENERAL_API_URL).href
    const response = await axios.post(url.href, { token });
```

File: src/Pages/Profile/MemberView/GetApiKeyModal.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    if (!apiResponse.error) {
+      setApiKey(apiResponse.responseData.apiKey);
+    } else {
+      bannerCallback(apiResponse.errror);
+    }
+  }
+
+  useEffect(() => {
+    if (!props.user.apiKey !== '') {
```
Comment (evanugarte) [inline src/Pages/Profile/MemberView/GetApiKeyModal.js]: what does this do? double negation?

File: api/main_endpoints/routes/Messages.js (hunk 8 above)
Comment (jonathanguven) [reply api/main_endpoints/routes/Messages.js]: done

File: src/Pages/Profile/MemberView/GetApiKeyModal.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+          <div className="sm:flex sm:flex-row-reverse">
+            <button
+              onClick={() => generateKey()}
+              className={'bg-green-600 btn inline-flex w-full justify-center rounded-md px-3 py-2 text-sm font-semibold text-white shadow-sm sm:ml-3 sm:w-auto'}                    >
+                    Generate Key
+            </button>
+            <form method="dialog">
+
+              <button className="bg-red-600 btn mt-3 inline-flex w-full justify-center rounded-md px-3 py-2 text-sm text-white shadow-sm ring-1 ring-inset sm:mt-0 sm:w-auto">
+                  Cancel
```
Comment (evanugarte) [inline src/Pages/Profile/MemberView/GetApiKeyModal.js]:
```suggestion
                  Close
```

**Review by evanugarte — APPROVED** [review]
